### PR TITLE
Add button to view raw logs during a build

### PR DESCRIPTION
**Pull request by manics**
Adds a `view raw` button to view the current build logs as plain text. Suggested by @bollwyvl in https://discourse.jupyter.org/t/binder-fails-to-compile-pygraphviz/9912/5 (at the end of the post).

This works by storing all log messages in an array, and creating a URL Blob object containing the logs when `view raw` is clicked. Note this won't update with new logs. You have to click `view logs` again.

Using `FakeBuild`:
image

Clicking on `view raw` opens a new tab:
image

I've also tested this in Chrome.

### CODE REVIEW

**Comment by bollwyvl** [comment]
Best feature I didn't have to write! It might also be possible to pull it directly out of the xterm, but it looks like it only stores the last 1000 lines, so this might be better.

It would be _slightly_ more ergonomic if it showed a useful filename, but wouldn't consider that a blocker, and it's apparently annoying to make it do the thing without actually pushing it to the DOM, which seems silly.

File: binderhub/static/js/index.js [hunk 1]
```diff
@@ -252,7 +253,20 @@ function setUpLog() {
     }
   };
 
+  $('#view-raw-logs').on('click', function(ev) {
+    const blob = new Blob([logMessages.join('')], { type: 'text/plain' });
+    this.href = window.URL.createObjectURL(blob);
+    // Prevent the toggle action from firing
+    ev.stopPropagation();
+  });
```
Comment (consideRatio) [inline binderhub/static/js/index.js:261]: Is this understanding correct?
1. Line 257: The blob constant contain all log messages concatenated into a string, and of a declared type of text/plain.
2. Line 258: An data based URL destination is via createObjectURL.
3. Line 258: The link named with `id=view-raw-logs` gets its `href` field automatically set to the data based URL.

I don't understand the following though:
- Line 259-260: Hmmm... This is a click event registered on a single link element, isn't it? Would that event propagate/bubble to the toggle button somehow? What confuses me is that the toggle button isn't a parent element of the `a` link that the `view-raw-logs` element is. I'm not sure how various click events can triggered registered functions in general, but my understanding of `stopPropegation` is to mark the click event as handled somehow which I guess it is and then perhaps that's a sensible practice - but unrelated to the toggle button.

**Comment by consideRatio** [comment]
Nice work @manics! What I think I understand LGTM, but I'm not sure about a not so critical part that I don't understand.

**Comment by manics** [comment]
Having a random `blob:` URL isn't great. An alternative is to force the file to be downloaded:
```diff
diff --git a/binderhub/static/js/index.js b/binderhub/static/js/index.js
index 289e392..70c160f 100755
--- a/binderhub/static/js/index.js
+++ b/binderhub/static/js/index.js
@@ -256,6 +256,7 @@ function setUpLog() {
   $('#view-raw-logs').on('click', function(ev) {
     const blob = new Blob([logMessages.join('')], { type: 'text/plain' });
     this.href = window.URL.createObjectURL(blob);
+    this.download = 'log.txt';
     // Prevent the toggle action from firing
     ev.stopPropagation();
   });
```
image

I can't decide what's better. What do you think?

File: binderhub/static/js/index.js (hunk 1 above)
Comment (manics) [reply binderhub/static/js/index.js:261]: The toggle function is attached to the header, not the button:
- https://github.com/jupyterhub/binderhub/blob/938324b95becd269da56decb4aff434af5fd9eda/binderhub/static/js/index.js#L255
- https://github.com/jupyterhub/binderhub/blob/938324b95becd269da56decb4aff434af5fd9eda/binderhub/templates/index.html#L149

**Comment by manics** [comment]
Just realised there's also a `binderhub/templates/loading.html` that needs to be updated

File: binderhub/static/js/index.js (hunk 1 above)
Comment (consideRatio) [reply binderhub/static/js/index.js:261]: Ah, thanks!

**Comment by consideRatio** [comment]
> I can't decide what's better. What do you think?

I'm used to opening an URL where i see a lot of blank text that I can select-all -> copy on. I think that because I'm used to that, like GitHub's "view raw" button. Was that one of the options via the blob url? Then I'm +1 for that!

**Comment by manics** [comment]
That's how it behaves at the moment- you'll get a new tab with a random in-browser blob URL.

**Review by consideRatio — APPROVED** [review]
This LGTM and make sense!

**Pull request merged**